File: translanguaging-IME/trie.py
import json
import heapq  # for efficient min heap operations

import heapq  # for efficient min heap operations

class TrieNode:
    def __init__(self):
        self.children = {}
        self.value = None


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, key, value):
        node = self.root
        for char in key:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.value = value

    def search(self, key):
        node = self.root
        for char in key:
            if char not in node.children:
                return None
            node = node.children[char]
        return node.value


def levenshtein_distance(s1, s2):
    if len(s1) < len(s2):
        return levenshtein_distance(s2, s1)

    if len(s2) == 0:
        return len(s1)

    previous_row = range(len(s2) + 1)

    for i, c1 in enumerate(s1):
        current_row = [i + 1]

        for j, c2 in enumerate(s2):
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)

            current_row.append(min(insertions, deletions, substitutions))

        previous_row = current_row

    return previous_row[-1]


def find_closest_matches(query, trie, k=5):
    min_heap = []

    def dfs(node, key_so_far):
        nonlocal min_heap
        if node.value is not None:
            distance = levenshtein_distance(query, key_so_far)
            heapq.heappush(min_heap, (distance, key_so_far, node.value))

    def traverse(node, key_so_far):
        dfs(node, key_so_far)
        for char, child_node in node.children.items():
            traverse(child_node, key_so_far + char)

    traverse(trie.root, "")

    return [(result[1], result[2], result[0]) for result in heapq.nsmallest(k, min_heap)]


# Example usage:
my_dict = json.loads(open("bopomofo_dict_with_frequency2.json", "r").read())
my_dict2 = json.loads(open("canjie_dict.json", "r").read())
trie = Trie()

# Insert keys into the trie
for key in my_dict:
    trie.insert(key, my_dict[key])
for key in my_dict2:
    trie.insert(key, my_dict2[key])

while True:
    query_string = ""
    query_string = input("Enter a query: ")

    closest_matches = find_closest_matches(query_string, trie, k=5)
    print("Closest 5 matches:")
    for match in closest_matches:
        print(f"Key: {match[1]}, Value: {match[0]}, Distance: {match[2]}")

File: IME extension/Trie.js
"use strict";

class TrieNode {
    constructor() {
        this.children = {};
        this.value = null;
    }
}
class Trie {
    constructor() {
        this.root = new TrieNode();
        this.keyStrokeCatch = {};
        this.type_list = new Set();
    }

    insert(key, value, type) {
        if (!this.type_list.has(type)) {
            this.type_list.add(type);
        }

        let node = this.root;
        for (let char of key) {
            if (!(char in node.children)) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        if (node.value === null) {
            node.value = value.map((element) => ({
                word: element[0],
                frequency: element[1],
                type: type
            }));
        } else {
            node.value = node.value.concat(value.map((element) => ({
                word: element[0],
                frequency: element[1]
            })));
        }
    }

    search(key) {
        let node = this.root;
        for (let char of key) {
            if (!(char in node.children)) {
                return null;
            }
            node = node.children[char];
        }
        return node.value;
    }

    /**
     * 
     * @param {string} query
     * @param {number} num_of_result 
     * @returns {array} array of objects of the form {distance, keySoFar, value}
     */
    findClosestMatches(query, num_of_result) {
        if (query in this.keyStrokeCatch) {
            return this.keyStrokeCatch[query];
        }
        let minHeap = [];

        function dfs(node, keySoFar) {
            if (node.value !== null) {
                let distance = levenshteinDistance(query, keySoFar);
                minHeap.push([distance, keySoFar, node.value]);
            }
        }
    
        function traverse(node, keySoFar) {
            dfs(node, keySoFar);
            for (let char in node.children) {
                traverse(node.children[char], keySoFar + char);
            }
        }

        function levenshteinDistance(s1, s2) {
            if (s1.length < s2.length) {
                return levenshteinDistance(s2, s1);
            }
        
            if (s2.length === 0) {
                return s1.length;
            }
        
            let previousRow = [...Array(s2.length + 1).keys()];
        
            for (let i = 0; i < s1.length; i++) {
                let currentRow = [i + 1];
        
                for (let j = 0; j < s2.length; j++) {
                    let insertions = previousRow[j + 1] + 1;
                    let deletions = currentRow[j] + 1;
                    let substitutions = previousRow[j] + (s1[i] !== s2[j]);
        
                    currentRow.push(Math.min(insertions, deletions, substitutions));
                }
        
                previousRow = currentRow;
            }
            return previousRow[previousRow.length - 1];
        }

        traverse(this.root, "");
        minHeap.sort((a, b) => a[0] - b[0]);

        const result = minHeap.slice(0, num_of_result).map(result => ({
            "distance": result[0],
            "keySoFar": result[1],
            "value": result[2]
        }));
        this.keyStrokeCatch[query] = result;

        return result
    }
}


async function build_Trie(){
    const trie = new Trie();
    const JSON_URLs = {
        ENGLISH: chrome.runtime.getURL('./src/english_dict_with_frequency.json'),
        BOPOMOFO: chrome.runtime.getURL('./src/bopomofo_dict_with_frequency2.json'),
        CANJIE: chrome.runtime.getURL('./src/canjie_dict_with_frequency.json'),
        PINYIN: chrome.runtime.getURL('./src/pinyin_dict.json'),
        THAI: chrome.runtime.getURL('./src/thai_dict.json'),
        VIETNAMESE: chrome.runtime.getURL('./src/vietnam_dict.json'),
    }

    for (let Language_key in JSON_URLs) {
        try {
            const response = await fetch(JSON_URLs[Language_key]);
            const dict = await response.json();
            for (let key in dict) {
                trie.insert(key, dict[key], Language_key);
            }
            // return trie;
        } catch (error) {
            console.error(`Error Building Trie from ${Language_key} JSON files:`, error);
        }
    }
    return trie;
}